llvm/lib/Target/EraVM/EraVMPostCodegenPrepare.cpp

```diff
@@ -0,0 +1,128 @@
+//===-- EraVMPostCodegenPrepare.cpp - EraVM Post CodeGenPrepare -*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass performs optimizations after CodeGenPrepare pass. Currently, it is
+// only performing transformation from signed to unsigned compares where it is
+// feasible.
+//
+//===----------------------------------------------------------------------===//
+
+#include "EraVM.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstrTypes.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "eravm-post-codegen-prepare"
+#define ERAVM_POST_CODEGEN_PREPARE_NAME                                        \
+  "EraVM optimizations after CodeGenPrepare pass"
+
+namespace {
+struct EraVMPostCodegenPrepare : public FunctionPass {
+public:
+  static char ID; // Pass ID
+  EraVMPostCodegenPrepare() : FunctionPass(ID) {
+    initializeEraVMPostCodegenPreparePass(*PassRegistry::getPassRegistry());
+  }
+  bool runOnFunction(Function &F) override;
+
+  StringRef getPassName() const override {
+    return ERAVM_POST_CODEGEN_PREPARE_NAME;
+  }
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    FunctionPass::getAnalysisUsage(AU);
+  }
+};
+} // end anonymous namespace
+
+// Generate unsigned compare instructions based on the predicates and extend
+// first operand if needed. If Pred2 is valid, generate second compare
+// instruction with binary operation Opc.
+static Value *generateUnsignedCmps(ICmpInst &Cmp, const APInt &Val,
+                                   CmpInst::Predicate Pred1,
+                                   CmpInst::Predicate Pred2,
+                                   Instruction::BinaryOps Opc) {
+  IRBuilder<> Builder(&Cmp);
+  assert(Val.getBitWidth() == 256 && "Expected 256-bit value for comparison");
+  assert(ICmpInst::isUnsigned(Pred1) && "Expected unsigned predicate in Pred1");
+
+  auto *Val256 = ConstantInt::get(Builder.getInt256Ty(), Val);
+  auto *Op256 = Builder.CreateZExt(Cmp.getOperand(0), Builder.getInt256Ty());
+  auto *NewCmp = Builder.CreateICmp(Pred1, Op256, Val256);
+
+  // Generate second compare instruction if needed.
+  if (Pred2 != CmpInst::BAD_ICMP_PREDICATE) {
+    auto *CmpVal = dyn_cast<ConstantInt>(Cmp.getOperand(1));
+    assert(CmpVal && "Expected constant operand for signed compare");
+    assert(ICmpInst::isUnsigned(Pred2) &&
+           "Expected unsigned predicate in Pred2");
+    auto *Cmp0 = Builder.CreateICmp(
+        Pred2, Op256, Builder.getInt(CmpVal->getValue().zext(256)));
+    NewCmp = Builder.CreateBinOp(Opc, NewCmp, Cmp0);
+  }
+  return NewCmp;
+}
+
+// Since we don't have signed compare instructions, try to transform them into
+// unsigned compares where possible. Currently, it is trying to do following
```

Comment:
It's always possible. But we limit this pass with constants in RHS and handle case of variables on DAG. (TODO: move it there - if you think it will improve maintainability).

Reply:
I don't think we want that, since ISEL can also generate signed compares. Yes, we can go ahead and disable all these ISEL optimizations, but it looks like an overkill to me since we would introduce a lot of local changes + we would need to worry about this every time we do the rebase to a newer version of the LLVM.